src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,269 @@
+  double res = discriminator.outputLayer.Forward(
+      std::move(boost::apply_visitor(
+          outputParameterVisitor,
+          discriminator.network.back())), std::move(currentTarget));
+  noise.imbue( [&]() { return noiseFunction();} );
+  generator.Forward(std::move(noise));
+  arma::mat temp = boost::apply_visitor(
+      outputParameterVisitor, generator.network.back());
+  temp.reshape(discriminator.predictors.n_rows, 1);
+  discriminator.predictors.col(numFunctions) = temp;
+  discriminator.Forward(std::move(discriminator.predictors.col(numFunctions)));
+  discriminator.responses(numFunctions) = 0;
+
+  currentTarget = discriminator.responses.unsafe_col(numFunctions);
+  res += discriminator.outputLayer.Forward(
+        std::move(boost::apply_visitor(
+            outputParameterVisitor,
+            discriminator.network.back())), std::move(currentTarget));
+  return res;
+}
+
+template<typename Model, typename InitializationRuleType, typename Noise>
+void GAN<Model, InitializationRuleType, Noise>::
+Gradient(const arma::mat& /*parameters*/, const size_t i, arma::mat& gradient)
+{
+  if (!reset)
+    Reset();
+
+  if (gradient.is_empty())
+  {
+    if (parameter.is_empty())
+    {
+      Reset();
+    }
+    gradient = arma::zeros<arma::mat>(parameter.n_elem, 1);
+  }
+  else
+  {
+    gradient.zeros();
+  }
+
+  if (noiseGradientDiscriminator.is_empty())
+  {
+    noiseGradientDiscriminator = arma::zeros<arma::mat>(
+        gradientDiscriminator.n_elem, 1);
+  }
+  else
+  {
+    noiseGradientDiscriminator.zeros();
+  }
+
+  gradientGenerator = arma::mat(gradient.memptr(),
+      generator.Parameters().n_elem, 1, false, false);
+
+  gradientDiscriminator = arma::mat(gradient.memptr() +
+      gradientGenerator.n_elem,
+      discriminator.Parameters().n_elem, 1, false, false);
+
+  // get the gradients of the discriminator
+  discriminator.Gradient(discriminator.parameter, i, gradientDiscriminator);
+
+  noise.imbue( [&]() { return noiseFunction();} );
+  generator.Forward(std::move(noise));
+  discriminator.predictors.col(numFunctions) = boost::apply_visitor(
+      outputParameterVisitor, generator.network.back());
+
+  discriminator.responses(numFunctions) = 0;
+  discriminator.Gradient(discriminator.parameter, numFunctions,
+      noiseGradientDiscriminator);
+  gradientDiscriminator += noiseGradientDiscriminator;
+
+  if (currentBatch % generatorUpdateStep == 0 && preTrainSize == 0)
+  {
+    // Minimize -log(D(G(noise)))
+    // pass the error from discriminator to generator
+    discriminator.responses(numFunctions) = 1;
+    discriminator.Gradient(discriminator.parameter, numFunctions,
+      noiseGradientDiscriminator);
+    generator.error = boost::apply_visitor(deltaVisitor,
+        discriminator.network[1]);
+    generator.currentInput = noise;
+    generator.Backward();
+    generator.ResetGradients(gradientGenerator);
+    generator.Gradient();
+    double multiplier = 1.5;
```

Comment:
I think this variable could be useful if we want to set different step sizes. In that case it is better to add an argument to the constructor. How do you think?